### DISCUSSION TITLE
Was Rust a mistake?

### DISCUSSION THREAD

**Discussion by Arcitec** [denoland/deno#10566] [General]
Looking at the issues with lack of mobile ports due to Rust compilation issues (see other issues about that), and the lack of bindings for various GUI libraries since those libraries would require Rust ports/wrappers, makes me wonder something in general for @ry:

- What are your opinions about the decision to choose Rust for this project?
- In hindsight, was it a better choice than C++?

I know that Rust enforces safer code and memory management, but it seems to hinder this project from getting ports of popular C++ GUI libraries. Kind of the same issue that Rust in general has.

**Comment by ry** [comment]
> What are your opinions about the decision to choose Rust for this project?

Rust has been fantastic for Deno. Zero regrets.

> In hindsight, was it a better choice than C++?

Absolutely. I'll never start a new C++ project. Rust services the same problem domain, but solves it better, in many ways. More than anything, having a single build system (cargo) and a system for linking in 3rd party code (crates.io) helps immensely. The biggest problem in C++ is linking to 3rd party code. Every project uses different systems: gn, gyp, cmake, autoconf, make.

 Rust is better in a myriad other ways too. But we really appreciate being able to pull in complex 3rd party native code without much trouble.

**Reply by MierenManz** [reply]
in your talk about deno you always say that you dislike centralized things like npm. But then why do you like crates.io alot?
Do you think that decentralized packages suit js better but not rust.
Or do you think that rust can benefit from a decentralized registry aswell?

**Reply by kitsonk** [reply]
Crates.io only impacts the building of Deno, not the usage of Deno.

Seems like you are here to pick a fight. 😢 if you don't like Deno, just don't use it.

**Reply by MierenManz** [reply]
oh no, that was not my intention. I was just curious, I'm not here to pick a fight
Sorry if it came over like that

**Reply by Arcitec** [reply]
@ry Hi Ryan, thank you very much for giving my slightly silly question such a great answer. You've actually sold me on the whole Rust idea. It's still a young language but I can definitely see how it cleans up the messy C++ ecosystem, and provides all the other Rust benefits like memory safety and a way cleaner syntax (not having 40 years of different legacy C++ syntaxes all bolted onto each other). I also appreciate how clean Rust's Traits are instead of C++'s classes. There are so many ways to shoot yourself in the foot or use wrong/outdated techniques in C++, so I can definitely see the safety enhancements gained by having Rust in Deno to ensure that random contributors have less ways to mess up the codebase. And perhaps it's also nice to not get 40 pages of template error messages (C++ errors) from the compiler anymore. ;-)

When you, a person I highly respect, said that "I'll never start a new C++ project, because Rust solves it better", you've actually given me the final push to start learning Rust. I've been eyeing the language but always felt like it was too small and had too few libraries. But the fact that it's been the number 1 rated language on [StackOverflow's "most loved languages" list](https://insights.stackoverflow.com/survey/2020#most-loved-dreaded-and-wanted) every year for half a decade now, by a huge margin, says something about how good it is.

I think that my worries about the future of Deno are going to be solved at the same rate that people are realizing how great Rust is and start creating/porting big libraries for it. Your message today, showing your support for Rust, contributed to helping me understand Rust's potential. And I recently saw news that they are starting to use Rust for the Linux Kernel.

It would be great to live 20 years into the future, and having a huge Rust ecosystem already. But I think that things like [rust-bindgen](https://github.com/rust-lang/rust-bindgen) and existing Rust libraries will tide me over and have solutions for *most* things...

So what I am saying is that I have decided to learn Rust and add myself to the list of users to help grow the language. It seems like it's starting to hit critical mass, when even the Linux Kernel decides that it's time to get rusty, and your words helped a lot. I don't think any other language in the world has more future potential than Rust, and I've kind of been itching for a reason to get involved.

Thanks for explaining the rationale behind using Rust for Deno, and what it brings to the project.

**Comment by lucacasonato** [comment]
Regarding GUI libraries: Rust has some fantastic up and coming libraries for GPU access (some written in pure Rust), most notably `wgpu` which we use for our internal bindings for the experimental WebGPU crate.

@crowlKats and @eliassjogreen are working on bindings to get WebGPU to render to a window. This should allow high performance, low level graphics with a modern API.

If you care more about higher level GUI bindings (something like electron), also consider http://deno.land/x/webview. This binds your system provided HTML engine into Deno as a UI framework.

**Reply by Arcitec** [reply]
Wow, thank you for showing me that. That [wgpu-rs](https://github.com/gfx-rs/wgpu-rs) library actually looks impressive. It's able to target WASM too. That would allow people to create custom 2D/3D GUIs, and that's definitely something an application developer would need.

I had seen the webview library but got the impression that it looked like a small and possibly unstable project. I don't base that on anything other than few GitHub stars and never having heard of webview before, though.

My main concern was actually the apparent lack of direct, native UI components. For example, [libui-node](https://github.com/parro-it/libui-node) or [React Native](https://reactnative.dev/). I wasn't able to find anything that could do that for Deno. And although a library like Electron/webview would suffice as an alternative, I just doubted the stability or maturity of webview. Perhaps I'm (hopefully) wrong about webview.

**Reply by lucacasonato** [reply]
Remember Deno has only been stable for a year. I think more libraries with more features will develop over time as the ecosystem grows in size.

**Reply by Arcitec** [reply]
That's true. I like everything about Deno's design much more than Node's. And Rust is a fantastic language. I also love that Deno can be easily extended with Rust modules, which means things like UI libraries will be able to be tightly integrated in the future. And the V8 engine itself is brilliant ([just-js](https://github.com/just-js/just) was able to integrate V8 JS with the Linux syscalls with almost as low overhead as native C/C++, thanks to V8's fast JIT). Hmm. I think Deno's future is incredibly bright.

**Comment by Symbitic** [comment]
I realize that the Deno maintainers have no regrets, but what about the wider ecosystem of developers? You need to know Rust to contribute to Deno. C/C++ remains very, very active. And popular. For all the love Rust gets on Stack Overflow surveys, its hype is somewhat bigger than its size (not my words).

It just doesn't seem fair to me that after more than a decade of developing Node.js core and extensions in C++, there's now no way for developers to contribute or extend Deno without learning a new language.

> I'll never start a new C++ project.

@ry This is more for curiosity's sake than any meaningful design decisions, but if C++ had a single tool for both pulling in dependencies and build system, would you be more inclined to (re)use it, or is it just the Rust syntax in general that makes it too good to give up?

**Reply by crowlKats** [reply]
> It just doesn't seem fair to me that after more than a decade of developing Node.js core and extensions in C++, there's now no way for developers to contribute or extend Deno without learning a new language.

well, people who know rust can contribute to Deno and dont need to learn C/C++ if they dont know it. this is an argument that can go both ways.

**Reply by kitsonk** [reply]
Rust memory safety wins hands down over C/C++. There are major reasons why there has been a significant move towards Rust, with statements by the like of v8 team stating if they had to do it all over again they would do it in Rust.

Personal opinion is that Deno being in Rust hasn't limited contributions.

**Reply by wwderw** [reply]
> It just doesn't seem **fair to me** that after more than a decade of developing Node.js core and extensions in C++, there's now no way for developers to contribute or extend Deno without learning a new language.

Not much is going to be.

I'm nowhere near the level as a lot of people on here (far from it).  I know a smattering of C++/GO/JS/GDScript/Python and that's really about it.  Enough to cause damage as my family would say.  There always seems to be something new to be learned, regardless if it's something totally new (as in a new language) or just new within a specific language that causes us to go back and change our code to reflect those new changes.

I think that we all have to be more forgiving with changes like that.  There is a saying that I think is appropriate for something like this: "Either change with the times or the times change you".  It sucks on occasion, but tis what it is.

**Reply by Symbitic** [reply]
Perhaps I should clarify what I meant. I am not asking/suggesting if Deno would be better off using C++. I am asking if there should be a way for C++ developers to use or extend Deno.

Some examples (not comprehensive):
1. Make the plugin/extension system more language-neutral.
2. Allow WebAssembly to be consumed at build time.

Please note that I have not expressed my personal opinions on Rust (or Deno for that matter). I just think that after all these years, developers who did work on Node deserve a bit more than "learn Rust or get lost." Shouldn't there be a way they can still contribute in some way?

**Reply by MierenManz** [reply]
> Make the plugin/extension system more language-neutral.
With the new plugin system that @eliassjogreen is working on it should be possible to use any language as a plugin if that language can compile to a cdylib

**Reply by ThatsNoMoon** [reply]
> I just think that after all these years, developers who did work on Node deserve a bit more than "learn Rust or get lost."

I don't think the Deno team owes anything to Node developers, other than perhaps gratitude for facilitating the demand for the type of project they're making. They definitely don't owe Node developers any compromises in their project design. Instead, Deno should allow people to use languages other than rust in plugins because *it would be better for Deno.* Saying Node developers "deserve" accommodation seems kind of entitled.

**Reply by vicary** [reply]
@Ruzie npm's problem is that it's a private company.

**Reply by kitsonk** [reply]
npm is owned by GitHub which is owned by Microsoft which is a publicly traded company. It is more appropriate to say that the npm registry is a closed platform with closed governance. It has been a desire by Deno to not force users to lock-in to a single closed ecosystem.

[npm in Deno](https://deno.land/manual@v1.20.3/node) is supported via a number of mechanisms currently and we are looking at ways of making it easier in the future.
